### ISSUE TITLE
Take advantage of Poolboy for connection pooling?

### ISSUE THREAD

**Issue by stevenproctor** [robconery/moebius#20]
I saw you were executing out to the `psql` command since you couldn't handle more than one connection  with the postgrex per a comment.

You might be able to take advantage of [Poolboy](https://github.com/devinus/poolboy) to be able to setup a connection pool instead of having to rely on an external system command.

**Comment by nurugger07** [comment]
This is a good call :+1:

**Comment by robconery** [comment]
The `psql` thing was mostly for bulk file execution I was doing in another project; we're using Postgrex to manage the lifecycle/spawn stuff BUT I do like the idea of having a process pool.

I'll look into this.

**Comment by robconery** [comment]
OK this is a touch over my head but I see the value in parallel execution. I'm wondering if OTP might be better suited? That's what I thought we were getting with `Postgres.Connection.connect` since it spawns a process - one for each query.

Am I wrong in my understanding of this? In other words, if 1000 queries come in, won't that spawn 1000 processes? We're not doing anything magical in the code - the time comes from the query I think...

**Comment by nurugger07** [comment]
I was going to look at this this evening. 

—
Thanks,

Johnny

**Comment by stevenproctor** [comment]
Poolboy is built atop OTP, so you get the process per connection, but what it gets you aside from having to recreate the connections each time, is that it allows the ability to create a work queue that would put an upper bound the connections allowed to a database.

While Postgres can handle many connections, it does have a limit, and we as system mantainers might not just want any application that tries to connect to the database to be able to take control of all the connections allowed by the database, especially with Erlang/Elixir based concurrency where 100k web connections might be happening at any time. ;)

**Comment by stevenproctor** [comment]
Thanks for thinking about this.  I look forward to seeing where you take moebius. :D

**Comment by robconery** [comment]
This is a dynamic equilibrium issue - Postgres will pool the connections no matter what and cap them. If you have 100k web connections then I doubt you'll settle for the (what is it... 10 concurrents?) the default with Postgres. I bring this up to thoroughly think on _what's a good idea_ vs. _what we need to do_.

I totally agree that Elixir/Erlang allows for a firehose to the DB. Once we crack the lid, however, we have a shitload of moving parts in OTP and Postgres connection management (as well as clustering).

I think @datachomp should weigh in on this, I know he's dealing with this first hand right now.

**Comment by stevenproctor** [comment]
I agree that it may just be an idea, and might not fit.  Was wanting to at least put it on the radar as something to think about if needed.

I would love to hear from @datachomp, both on this and more about what he is doing/finding in general. ;)

**Comment by robconery** [comment]
Hopefully don't read this as pushback necessarily; I just want to know what's going on is all :).

**Comment by nurugger07** [comment]
The benefit of using poolboy is that it manages a collection of open connections. When we request a process id from the pool, it's checked out of the collection and unavailable until the transaction is finished. Then it's checked back in and available to be reused. The big gain is that we don't incur the penalty of opening connections, they are just available.

**Comment by robconery** [comment]
Right - _that_ is the question: what is the penalty? Or, _is_ there a penalty for opening a connection? If that penalty is only demonstrable in the 5000 concurrent range, I don't think I'd be swayed. If it can be significant in the 500 range... yes. What I'm wondering is "does it stack?" and if it does, then yes we need Poolboy. If it doesn't stack then I don't care.

I dig the pooling concept - but what we're talking about is exactly as @nurugger07 put it: **the cost of opening the connection**. If that cost remains the same through 5000 simultaneous connections we haven't gained anything.

I think. There's a ton of room for me to be wrong.

**Comment by stevenproctor** [comment]
@robconery Pushback is fine; every line of code adds extra overhead, and if a feature is not needed, or isn't worth the cost (implementation as well as ongoing maintenance), then it should be questioned. ;)

I don't know the answers to any of these off hand.  My goal of opening this issue was more to raise the awareness of Poolboy, to see if it might work for connection pooling instead of _having_ to roll your own pool management, if connection pooling was needed.

As I also don't know the extent of what your plans/goals are with moebuis, as far as how much you are wanting this to be a library that gets picked up and used by the community, or if this is more of a proof of concept only, so I would fully expect that to play into your decision on if this should be investigated or not.

Either way, thank you for this project, it will be interesting to see where this goes.

**Comment by robconery** [comment]
Cheers! Thanks @stevenproctor appreciate the response. It's out there, and I don't think I can shy away from its use. I'm all for improving so let's focus on that, especially perf!

I had a great convo with @datachomp the other day that Postgrex is so efficient it "makes PostgreSQL the bottleneck". Which I can see; Erlang and all.

To me, the question is _at what point_ does a live connection pool add value? I don't know honestly and I think we'll need to try and break it to see what happens. I'm leaning toward thinking that there are a number of other things un-code-related that can help. Specifically:
- bumping your concurrent connections in PostgreSQL 
- upping your cores? I know I wouldn't want to suggest this as an alternative but I'd love to see what happens with additional cores. The idea here is that if your app is going bonkers, adding a core or two might solve a _whole shitload_ of issues - Poolboy might be inconsequential
- upping RAM (same as a above) simply to hold ops in memory

I guess the one thing we can do right now is see where it breaks? I'd be all for another issue RE perf testing this thing.

**Comment by xivSolutions** [comment]
If you run something like this:

``` elixir
  def people(qty) do
    Enum.map(1..qty, &(
      [
        first_name: "FirstName #{&1}", 
        last_name: "LastName #{&1}", 
        address: "666 SW Pine St.", 
        city: "Portland", 
        state: "OR", 
        zip: "97209" ]))
  end

  def write_benchmark(qty) do
    people(qty)
    |> Enum.map(fn(p) -> db(:people) |> insert(p) end) 
  end
```

with an input of 100 +/- a few (give or take a few, possibly machine dependent), you get the "too many connections" error with the default Postgres installation. Possibly I'm missing something, but if you pull Moebius into an application and want to do the FP equivalent of iterating over a list of records to `insert`, things fall over fast. 

Am I missing something about how this should be done in Elixir, or is the the direct expression of the problem discussed above?

**Comment by robconery** [comment]
Yeah I was literally thinking about this when I was on a ride today. I don't think we're cutting the process off when the query is finished. I'll get an issue on that...

**Comment by mlakewood** [comment]
So the main advantage of connection pooling is speed of query. If your queries are trivial and fast (like all good queries should be) then actually a significant portion of the request time is setting up the connection in the first place (I dont have exact number on this, but thats my understanding from other languages/platforms). 
This same concept plays out in other realms as well. If you have lots of little js files, you are better concatenating them and pulling them all in one request because then you dont have all the wasted http handshake and setup time over and over again. A database connection is roughly the same. If you keep a set of connections open, then you can make individual queries much faster.

**Comment by robconery** [comment]
@mlakewood That's the claim I keep hearing :) and yes, I do believe it. However I don't have any hard data on this, nor do I think it would make _any difference_. Let me explain.

If we have a query that executes in 0.05s and a connection that takes 0.05 seconds to open, we have a total execution time of 0.1 seconds. That's as fast as you blink. Now, we could speed that up so that in instead of 0.05 seconds to open the connection we drop it to 0.001 seconds (almost instant) for pulling a pooled object.

Now we have a total execution time that is marginally faster in terms of perception - yet I will grant that's a nice boost. This is where I stop and weigh the gains and costs of this.

In a normal (non-Erlang) system, this gain would have a ripple effect and make everything else that much tighter and faster. Given that we're using Erlang and spawns/workers which execute in parallel - that gain isn't realized since the ripple effect doesn't exist. The problem that _does_ exist is fire hosing Postgres - which I've successfully done :).

In short: 1 query will execute in 0.1 seconds. 100 queries at the same time will execute in 0.1 seconds - parallel processing.

Now the cost of this is mostly on the dev side - maintaining a proper pool with locks/checkins and handling transactions properly (locking up a pooled PID) and long-running queries doing the same and I have to wonder if this is a rats nest I don't want to climb into (messed up metaphor)? At it's core this is caching - and **I will not gleefully walk into a caching nightmare** to shave a few milliseconds :).

Also - as @datachomp has pointed out to me a few times: _never let your app do the pooling_. There are things you can drop in front of Postgres to manage the pools properly when your DB gets to that size.

Anyway - long response - if you have a reply (which I hope you do) please make sure to spell out the parallel aspect of the gain.

Cheers

**Comment by mlakewood** [comment]
So I agree that due to erlang the pooling on the client side is less of a problem (ie not so many ripple effects). However it costs Postgres time and CPU cycles to open and close a connection. IMHO the less work the db has to do the better. 

But I also agree that you can implement the pooling use pgbouncer etc. But from an administration POV this is another moving part to deal with.

I guess what i'm saying is from an efficiency POV, if you dont have to do the work at run-time why would you?

**Comment by robconery** [comment]
How does it cost Postgres "time and CPU cycles" to open a connection? And, if so, what does that amount to? I have to think that if Postgres slowing down is our problem... that's a good problem to have. I think I might disagree with your assertion about offloading DB work; it's probably the **least utilized** part of the entire stack.

For instance we can write 10K records to Postgres in about 0.6 seconds. 10K. Postgres can handle this and a lot more just fine. That doesn't bother me.

I'm not sure I follow your last question. What does a connection pool have to do with "efficiency". I'll counter this notion with "no code is the best code"; especially when it comes to my free time and walking face-first into one of the most well-known problems in computer science.

I'll swallow that extra 0.000001 second Postgres has to suffer through to respond.

**Comment by mlakewood** [comment]
There is a handshake that has to happen to open the connection, so Postgres has to be part of that. Which means that it is doing work. Both ends need to do some work.

Also I think when you frame the issue as how long one request is taking I think you miss the fact that as more requests are using the same connection, you no longer have to do any of that work. So in the case of 1000 requests, if a connection takes 0.05 seconds (from above indicative of magnitude not actual numbers) \* 1000 thats actually a lot of work that you dont have to do. 

I would also disagree that the database is about the least utilized part of the stack. I think it can be, but once you put multiple servers in front of it serving  a reasonable amount of traffic for each, that adds up pretty quickly. Postgres is inherently (effectively) threadbound and blocking. You cant get away from that. So you need to ensure that each operation takes the least amount of time to do. And protect the database, because if you lock it up for some reason, all those services die.

There is also the question of bulkheading. Ie, if this service maxes out its connection limit (for whatever reason) then it wont keep making connections. This can stop it from interfering with other services that are also talking to the db. I can conceive of a situation where a single service could chew up connections one after another, not doing anything and not releasing them and so DoS all the other services talking to postgres. Now obviously this is a bug, but defence in depth is a handy thing to have.

I think it basically comes down to a question (like a lot of things is software) of scale. At a certain scale I dont think pooling matters. At other scales it really does matter.

If the effort to pull in poolboy is not that great, and you want this library to be used at higher end scales then I would think you should do that. If it isnt and you dont, then dont.

Hopefully we have all learned from the discussion :)

**Comment by robconery** [comment]
@mlakewood Thank you for a really detailed response - much appreciated. Before I go on: this is tough stuff and I can't tell you how much I value responses like this :). Thank you :).

First: yes we both agree that _work is going to happen_. I'm not refuting that.

>  I think when you frame the issue as how long one request is taking I think you miss the fact that as more requests are using the same connection, you no longer have to do any of that work

This is a double issue. Yes, I completely understand the idea of reusing an open connection - you abate the cost of reopening the thing that was once opened. _I agree_. I think your supporting calculation doesn't apply, however. There's no scorecard here, there's no _work tally_. I do agree that _something happens_ but I couldn't care less that it happens. Your point about the accretion of work is indeed valid, but that's what I addressed in my last paragraph above: _if we get to that point Moebius will not be the solution_.

I would never rely on an ORM/DB tool to handle connections properly if we're handling 1000 requests/second. At some point you need to hand off management of your database and if we're doing that volume - @datachomp can have it (or someone of his ilk).

> once you put multiple servers in front of it serving a reasonable amount of traffic for each, that adds up pretty quickly

At this point your argument is turning into one of scaling. I'll counter with what I said before: **this is Erlang**. I can do with Erlang what typical web stacks (.net, rails, etc) can do with 10 servers. I'll also repeat my point that _this is our problem_. I can DDoS my Postgres server quite easily - I've done it. Putting multiple servers in front of Postgres with an Erlang/Elixir front end would be catastrophic if the traffic warranted. At that point if I don't have a fleet of DBAs I'm an idiot - and again _Moebius connection pooling will not be my solution_.

> if this service maxes out its connection limit (for whatever reason) then it wont keep making connections. This can stop it from interfering with other services that are also talking to the db. I can conceive of a situation where a single service could chew up connections one after another, not doing anything and not releasing them and so DoS all the other services talking to postgres. Now obviously this is a bug, but defence in depth is a handy thing to have.

This problem is precisely what we're working on at the moment - and it's not something a pool can solve. The only thing that addresses a MAX_POOL problem is brokering, which is what @nurugger07 is spiking right now in one of our branches. Also: yes, anyone can write code that blows their DB up (I've done it) - I can't solve bad design :).

> I think it basically comes down to a question (like a lot of things is software) of scale. At a certain scale I dont think pooling matters. At other scales it really does matter... If the effort to pull in poolboy is not that great, and you want this library to be used at higher end scales then I would think you should do that. If it isnt and you dont, then don't.

This reads, to me, like a "shrug and walk away" - basically suggesting "if scaling doesn't matter to you then... whatever". I don't want to put words in your mouth - but this is how it reads.

I would suggest to you that you're looking at this problem from a classical perspective. One that holds to "connections are expensive so let's optimize for expensive connections". I don't have numbers on how expensive connections are in Elixir but I'll just agree for now - sure they're expensive in terms of time.

If I bump `postgres.conf` to 1000 concurrent connections and `spawn` every process in my app that wants to talk to my database - _when does that become an issue_.

This is a "Dynamic Equilibrium" - which means there are two processes operating on the equation at the same time: Load and Execution. With Elixir I can load Postgres quite quickly. With Postgres I can open the funnel and allow more queries in - pressing my CPU/RAM and working Postgres to the point of a lathery sweat. The _funnel_ - the limiting factor here - is the **size of the pool**. In some ways if a connection takes longer to spin up then it's kind of a good thing in the same way that a sail boat dragging a rope in a storm keeps it upright.

My point with this is that _scale is all I'm thinking about right now_. I'm not being casual about it, not suggesting it's someone else's problem Elixir is so damned fast and the parallel processing approach so wonderfully efficient that I'm actually worrying about _destroying databases_. The last thing I think I need is a faster connection.

Of course, I could be completely wrong. But from what I've seen with the tests in front of me: _Moebius isn't the issue_. Postgres maxing its connections are the issue. More than that - PG's ability to handle the load when I open the connection limit is also the problem. _These are literally our failure points_ in our scaling tests: **the database**.

Maybe you can see why I'm not paying attention to pooling :).

**Comment by stevenproctor** [comment]
Some questions that I feel might be missing from this to be answered as well, to help determine if a connection pool might be of use or not.

Again, I mainly opened this pull request as food for thought, and opening up Poolboy as a possible option to use if a connection pool is needed.

These questions are mainly to help flesh out the use case of moebius, and then be able to express that to people who might be interested in using it so that they can know if their needs might align.

_**1) Is there an expectation of of a tool like pgbouncer to be used in conjunction with this?**_

If the answer is yes, then it may just need to be more explicit as part of the README.md or somewhere else even.

_**2) What is the expected deployment structure of the Elixir app to be used and the PostgreSQL database?**_

This comes into questioning how much network latency plays in with establishing new connections.  Some protocols for communication are cheap, where other protocols require multiple round trip handshakes to agree to a connection being established, and what does a secure/encrypted connection look like compared with that as well? 

Are things expected to be run on the same box? In the same datacenter? Same private network but across data centers?  The latency cost of establishing a connection is another cost factor to consider if/when some tests are run to find the additional cost of setting up a connection, and not just testing it on the same device for the cost of a new connection.

Maybe @datachomp has an understanding of what this would look like and the impact it might have if higher latency.

_**3) What is the vision of the supervision and failure recovery strategy of connection failures?**_

As a consumer of this library, should I expect the connection processes to be supervised?

How would someone distinguish between some number of connection processes that are failing and dying only because we are creating too many and if we slow down we are good (the use case for `pgbouncer`), as opposed to a more systemic database issue?

With a smaller number of connections that is set to be "safely" "under capacity" (like a connection pool would likely be configured with), if a number of processes fail in a given set of time, there is a higher likelihood that something going on with the database.

On the flip side, if `pgbouncer` is killing a connection process because we "occasionally" do "one process too many", from a supervision and recovery/restart perspective, is there a way that is expected to be determined?

Is 10 failures in 10 milliseconds from a connection pool of size 10 thought of and treated in the same manner as 10 failures in 10 milliseconds from 10_000 connections because `pgbouncer` is killing them off?

Are the connections done with something like `pgbouncer` expected to not be supervised, but just linked processes to the process that is running the query?

If the connection processes are under a supervisor and it determines a restart is needed as configured, and it kills off the other children it is supervising, it is okay that if it will tries to kill all 10_000 processes, and potentially all other (request processing???) processes that are using them if they are linked as well?

Is supervision not expected to be used if a connection pool is not setup, and `pgbouncer` is the expected solution?

A consumer of this would need to know because that directly impacts their process structure and how they need to setup their supervision and restart strategies. 

_**3b) As a bonus to this (because I would guess it is not on the roadmap, and may not ever be), depending on the supervision strategy could there be a smartness to knowing if read queries vs writes are happening?**_

This is could play in with replication of a PostgreSQL instance that reads could have a different supervision hierarchy to get a connection (pooled or not) from, that would be able to tell if a given server is overloaded and restart to point to a different configured replica, or even spread out the connections across a given set of replicas to be able to read from.

This could play out depending on if there are different connections allowed to be configured (e.g. CQRS, or reporting database for some aspects of the app) where it would be safe to _read_ from another replica, but maybe the writes don't go through.

There could be a use case where connections could either be round robin'ed, or always from the same replica with a new replica being picked if the current one becomes "unavailable"/unresponsive, e.g. network partition, server overload of a different query, etc.

This premise of this bonus question is thinking about the supervision strategy could make these options either readily available or not depending on how the supervision and error recovery is built.

__Take or leave each of these individually or all together__

This is all food for thought, and questions that might spark some other ideas or be dismissed.

@robconery, this is your baby, and I thank you for sharing with the world, and please recognize that anything I say is as an observer and is intended as food for thought.

This is an interesting looking project, and while I don't know if I will get to use it in anger any time soon,  it is on my radar to watch develop.  I am happy to help discuss some trade offs of supervision if needed as well, if you want someone in addition to @nurugger07 to discuss some of the trade-offs.

**Comment by robconery** [comment]
I should add that if you're going to tell me that PG is dying because it plays a part in opening a connection, I'll want some numbers to go along with that. Right now I'm going with the idea that open connections are more of a problem then opening the connections themselves:
http://hans.io/blog/2014/02/19/postgresql_connection/

**Comment by stevenproctor** [comment]
As a pure aside, this is fascinating discussion, as it helps me see some of the things in the PostgreSQL ecosystem I was not aware of. :grinning:

**Comment by robconery** [comment]
@stevenproctor This is a lovely discussion on all fronts - thanks for raising it :). To be ultra clear: **no hackles are raised** on my part. I've never made a database tool for a platform/runtime/language that is this fast, distributed and elegant so this conversation _very much needs to happen_. So thanks :).

> Is there an expectation of of a tool like pgbouncer to be used in conjunction with this?

No. Not at all. pg_bouncer was brought up only because of a conversation I had with @datachomp about managing connections at scale. His point was that he'd never trust a db tool to manage connections properly... _at scale_. For 99% of the time I think we're OK.

> What is the expected deployment structure of the Elixir app to be used and the PostgreSQL database?

This isn't something I'm prepared to deal with in the tool itself; even with a connection pool the processing time is marginally influenced. 100 connections over a network in a distributed system are exactly as slow as one. A connection pool shaves a marginal hair here (of course, depending on the latency in question).

> As a consumer of this library, should I expect the connection processes to be supervised?

That's one of the problems I'm trying to avoid. A dropped connection is useless - especially in a transaction. One thing I just don't want to deal with. Open late, close early is my mantra for this reason and I think it holds here.

> could there be a smartness to knowing if read queries vs writes are happening?

Absolutely but then we get into the notion of _brokering_ - which connections go where and why. For transactions, I think a dedicated connection is paramount. Same with long-running queries. If we did use a pool (which we have to if we're brokering) then we would want to discriminate. Just had this conversation the other day :).

The questions you raise @stevenproctor open the door to everything I'm trying to avoid by stepping too deeply in it with connections. Once you manage this one thing, you manage it all and then this becomes the number one Issue topic - which is _precisely why I'm pushing back so hard_:
- I don't see the necessity for managing these things aside from hair-splitting time differences (even at scale) and
- I frankly don't want to support this into the future. I had a long talk with @josevalim about this the other day on IRC. It's a headache. I think I want to see a problem before I think of a solution TBH.

Again: _thank you_. This is a difficult discourse and I respect each of your opinions. Let's stay open and thoughtful even in disagreement. I really do look forward to your replies because **I will be the first to admit if I've got this all wrong** and we'll be the better for it!

**Comment by stevenproctor** [comment]
@robconery I salute you!

As I told @josevalim when I was asking him about deciding to create Elixir, it takes a special kind of crazy to put something out to a wide unknown audience and then have to support that in perpetuity, with anyone/everyone coming and saying how it should be done. :wink:

Cheers! :beers:   ( :house:-brewed of course...)

**Comment by robconery** [comment]
@stevenproctor You brew! What style! Oh jeez I should just invite you to slack... email?

**Comment by stevenproctor** [comment]
I don't brew :frowning: (yet!) but would love to learn more details about it. :grinning: 

I do know of your love of it from following you though.

**Comment by josevalim** [comment]
> As a consumer of this library, should I expect the connection processes to
> be supervised?
> 
> That's one of the problems I'm trying to avoid. A dropped connection is
> useless - especially in a transaction. One thing I just don't want to deal
> with. Open late, close early is my mantra for this reason and I think it
> holds here.
> 
> Let me just intervene here. Restarts are not the only benefit of
> supervision. In fact, it is completely sensible to supervise your
> connections even if you are not restarting them when they crash because:
- when your app is shutdown, the supervisor will take care of shutting down
  all connections properly
- when using observer, you will be able to introspect the supervisor to see
  how many connections you have opened and introspect the connections
  themselves
- supervisors will also allow you to do hot code swapping

Generally the question of "should I supervise those processes" is always
yes, even if you don't want restarts, and this case is no exception. It
doesn't mean you need to pool, it just means you want to start connections
as a child of a supervisor instead of directly calling start_link.

## 

_José Valimwww.plataformatec.com.br
http://www.plataformatec.com.br/Founder and Director of R&D_

**Comment by xivSolutions** [comment]
Jumping into the middle of a discussion between some heavyweights...

My strongly felt/loosely-held position is that scaling, and how it is managed (in this case, related to pooling/brokering/db access) is an application issue, as the implementaiton needs of applications will differ. 

Connection pooling and brokering are non-trivial, and by building them into Moebius, we would be asking for big headaches in effort to assuage concerns of a minority of use-cases. As @robconery and @datachomp suggest, building the connection pool/management into the query tool itself immediately imposes constraints. 

All that said, I am left feeling like the responsible approach would be to make sure there are some hooks available, when needed, such that pooling/brokering can be used if needed, without constraining which pool/broker, etc. Maybe, if @nurugger07 's broker and/or some other pool mechanism are in fact deemed necessary in the default implementation, then set those up as defaults. 

I think it's important to design with "scale" in mind as a potential factor, but also important not to base the entire design on the 10% use-case scenario. 

Great discussion here, and kept nicely civil while also expressing strong views! Following along, and working my ass off to absorb the interesting issues presented by Elixir/Erlang concurrency and Db access. Almost the inverse of the normal problem....

**Comment by mlakewood** [comment]
Thanks all for the civil discussion! Im also very open to being wrong and think this is a super valuable conversation to have.

To address the last part of my comment. I was trying to be really careful around my wording. My attempt was to say that I think there is great value in deciding what you want the library to do and what you don't. Taken with the idea that connection pooling somewhere is necessary at a certain scale then one valid decision is to say that moebius fills a certain part of the market that doesnt want to do connection pooling inside the elixir app. If that is the route you take then spelling that out (And perhaps has proctor suggests pointing people towards pgbouncer et al) in the readme is a good move. My phrasing was meant to convey that this decision is entirely up to you and which ever way you decide is fine.

In that frame I would also agree (with some caveats) with multiple of @xivSolutions comments. 
1. I think one possible approach if you arent going to build in poolboy is to make it relatively easy to compose in. Or as mentioned about point to other solutions such as pgbouncer.

For some extra data around the discussion.

A few useful links around the cost of connections in Postgres.

http://hans.io/blog/2014/02/19/postgresql_connection/
https://wiki.postgresql.org/wiki/Number_Of_Database_Connections

I think you are right that there isn't much direct benefit to the elixir app around connection pooling. I think there is benefit to postgres around connection pooling. And I think we both agree on that (correct me if im wrong). So it comes down to where you want to solve the issue. At that point in my mind there isnt a hard and fast rule, so its going to be up to you (and the consumers of your library) about where they want to solve it.

My preference is for in-app because I have more control and flexibility. Others would want to use pgbouncer. In my mind both are valid decisions with the right data input. These are the hard decisions of software :)

Thanks again for the discussion. Its always hard to get nuance and clear communication across in text format, and I really appreciate the open mind and open communication that this thread has had.

EDIT: as I just saw we linked to the same article about pg connections.
Your take on that article has certainly shifted my idea about the cost with opening vs leaving open. I havent fully digested that outcome. But wanted to comment here to make sure it didnt come across as "I didnt read what you linked to". I'll add a new comment when I've digested.

**Comment by mlakewood** [comment]
So I think that although "opening connections" and "open connections" are not exactly the same and have different effects, they do have some things in common.

So I think through this discussion I have learnt and moved my position. I think the issue is fundamentally about how you deal with an (almost) unbounded resource talking to a bounded resource. Elixir/erlang have such a high level of concurrency that they can be thought about as unbounded, where as postgres has a very set amount of resource (something like connections =  (cores \* 2) + number of spindles). This is similar to other ecosystems where you have an async webserver and Postgres, except most likely more extreme.

In this more abstract context I think its less about how much opening a connection costs, and only by relation how much a open connection costs (ie the reason you can only have so many connections open is because they cost something) and more about managing a known finite resource. A process pool (and this is essentially what poolboy et al are, rather than a connection pool like in java land) is one way of solving this issue. And I think pgbouncer is another way of doing that. And I think the only thing that moebius needs to do is decide if it wants to take responsibility for that, or to decide to not concentrate on that problem.

So I think although I have more information and a new view on what the issue is (and I hope a better understanding now) the conclusion that I've come to hasnt :)

Thanks again for the conversation.

**Comment by fishcakez** [comment]
Ecto style pooling/transactions has been moved outside Ecto: https://github.com/fishcakez/db_connection. If any of you would like to help port postgrex to use it that would be great.

**Comment by mlakewood** [comment]
Oh interesting. Ideas behind the reasoning of doing that? Was it simply, lets decouple this for code reasons? or some other more operational reason.

**Comment by fishcakez** [comment]
The main reason is that next version of Ecto is requiring a certain way to handle sockets and all driver maintainers will need to change their driver quite a bit. Moving that code into a separate project meant that code needs to be written once. After making that decision there are ample opportunities to make other improvements/clean up code that only need to be implement once.

For example to do a transaction on Ecto 1.0 using poolboy it requires at least 5 GenServer calls:
- Checkout worker from poolboy
- Checkout connection from worker
- Begin query
- 1 call per query
- Commit query

With DBConnection.Poolboy it requires 2 GenServer calls:
- Checkout connection from poolboy
- Checkout socket from connection (and do any number of queries without making a call)

In this case all the transaction logic and state is in one module and one process, which meant things could be cleaned up a lot too. Whereas in Ecto 1.0 it is spread out over multiple modules and three processes.

**Comment by mlakewood** [comment]
Oh cool. that makes a lot of sense. I was wondering how transactions worked in Ecto, or at least how we should be doing them as a user. Phoenix doesnt have a lot of info on how it thinks it should be done and points at Ecto for guidance around this, which really just has the one function. When will all this land in an Ecto release?

**Comment by nurugger07** [comment]
@fishcakez as per our conversation on issue #37 I would be happy to help out, even if it's just testing on our end. Would you like to chat some time to discuss possible action items?

**Comment by fishcakez** [comment]
Yes we can do that definitely. I would appreciate it very much.

**Comment by nurugger07** [comment]
@fishcakez awesome! Want to email me & we can set up a time to talk?

**Comment by fishcakez** [comment]
Email is great as I am a bit uncertain about availavility until the weekend.

**Comment by nurugger07** [comment]
@fishcakez sure, just email me and we can start the conversation

**Comment by robconery** [comment]
This is done and in master now.
